src/main/java/com/rabbitmq/client/impl/AMQCommand.java

```diff
@@ -101,17 +101,22 @@ public void transmit(AMQChannel channel) throws IOException {
 
         synchronized (assembler) {
             Method m = this.assembler.getMethod();
-            connection.writeFrame(m.toFrame(channelNumber));
             if (m.hasContent()) {
                 byte[] body = this.assembler.getContentBody();
 
-                connection.writeFrame(this.assembler.getContentHeader()
-                        .toFrame(channelNumber, body.length));
+                Frame headerFrame = this.assembler.getContentHeader().toFrame(channelNumber, body.length);
 
                 int frameMax = connection.getFrameMax();
                 int bodyPayloadMax = (frameMax == 0) ? body.length : frameMax
                         - EMPTY_FRAME_SIZE;
 
+                if (headerFrame.size() > frameMax) {
+                    throw new IllegalArgumentException("Content headers exceeded max frame size: " +
+                            headerFrame.size() + " > " + frameMax);
+                }
+                connection.writeFrame(m.toFrame(channelNumber));
```

Comment:
moved this code down to after size check to avoid leaving channel in invalid state in case exception was thrown